public/scripts/d3/d3.js

```diff
@@ -1 +1,130 @@
-//  Add d3 functions here
+d3.json(echo.margins.jsonUrl, (err, data) => {
+  data.sort((a, b) => d3.descending(a.avgRating, b.avgRating));
+
+  let uniqueTags = [];
+  data.forEach((d) => {
+    uniqueTags.push(d.tag);
+  });
+  uniqueTags = uniqueTags.filter(onlyUnique);
+
+  const nodePosition = [];
+  const startingCx = 160;
+  let startingCy = 120;
+  const sortedByTag = {};
+
+  uniqueTags.forEach((t) => {
```

Comment:
Okay, this is confusing, lots of mutation, two different things happening at once in this loop.

You want to bin the nodes by tag:

```js
const nodesByTag = nodes.reduce((nodesByTag, node) => {
  const tag = node.tag
  const bin = nodesByTag[tag]
  if (bin === undefined) {
    nodesByTag[tag] = [node]
  } else {
    nodeByTag[tag].push(node)
  }
}, {})
```
gives us
```
{
  'cat': [{}, {}],
  'friend': [{}],
 ...
}
```

This is much cleaner, do you see why. Lots and lots of variables floating around finding it hard to work out what is going on